File: src/cli/roadmap/list.ts
import type { Command } from "commander";
import { getClient } from "../../lib/client.ts";
import { printError, printPaginated, resolvePageSize } from "../../lib/output.ts";

export function registerList(roadmap: Command): void {
  roadmap
    .command("list")
    .description("List roadmaps")
    .option("--limit <n>", "Limit results")
    .option("--cursor <token>", "Pagination cursor for next page")
    .action(async (opts: { limit?: string; cursor?: string }) => {
      try {
        const client = getClient();
        const results = await client.roadmaps({
          first: resolvePageSize(opts),
          after: opts.cursor,
        });
        const items = await Promise.all(
          results.nodes.map(async (r) => {
            const owner = await r.owner;
            return {
              id: r.id,
              slugId: r.slugId,
              url: r.url,
              name: r.name,
              description: r.description,
              owner: owner ? owner.name : null,
            };
          }),
        );
        printPaginated(items, results.pageInfo);
      } catch (err) {
        printError(err instanceof Error ? err.message : "List failed");
      }
    });
}

File: internal/cli/roadmap/list.go
package roadmap

import (
	"context"

	"github.com/spf13/cobra"

	"github.com/shhac/lin/internal/linear"
	"github.com/shhac/lin/internal/output"
	"github.com/shhac/lin/internal/ptr"
)

func registerList(roadmap *cobra.Command) {
	var limit string
	var cursor string

	cmd := &cobra.Command{
		Use:   "list",
		Short: "List roadmaps",
		Args:  cobra.NoArgs,
		Run: func(_ *cobra.Command, _ []string) {
			client := linear.GetClient()
			pageSize := output.ResolvePageSize(limit)

			after := output.ResolveCursor(cursor)

			resp, err := linear.RoadmapList(context.Background(), client, pageSize, after)
			if err != nil {
				output.PrintError(err.Error())
			}

			items := make([]map[string]any, len(resp.Roadmaps.Nodes))
			for i, r := range resp.Roadmaps.Nodes {
				var ownerName *string
				if r.Owner != nil {
					ownerName = &r.Owner.Name
				}
				items[i] = map[string]any{
					"id":          r.Id,
					"slugId":      r.SlugId,
					"url":         r.Url,
					"name":        r.Name,
					"description": r.Description,
					"owner":       ownerName,
				}
			}

			pi := resp.Roadmaps.PageInfo
			output.PrintPaginated(items, &output.Pagination{
				HasMore:    pi.HasNextPage,
				NextCursor: ptr.Deref(pi.EndCursor),
			})
		},
	}

	cmd.Flags().StringVar(&limit, "limit", "", "Limit results")
	cmd.Flags().StringVar(&cursor, "cursor", "", "Pagination cursor for next page")
	roadmap.AddCommand(cmd)
}

